File: cpp/239_sliding_window_maximum.cpp
#include <deque>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    deque<int> max_indices;
    vector<int> result;
    result.reserve(nums.size() - k + 1);
    for (int end = 0; end < nums.size(); end++) {
      int start = end - k + 1;
      if (!max_indices.empty() && max_indices.front() < start) {
        max_indices.pop_front();
      }
      while (!max_indices.empty() && nums[max_indices.back()] < nums[end]) {
        max_indices.pop_back();
      }
      max_indices.push_back(end);
      if (start >= 0) {
        result.push_back(nums[max_indices.front()]);
      }
    }
    return result;
  }
};

File: rust/239_sliding_window_maximum.rs
impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let k = k as usize;
        if k == 1 {
            return nums;
        }

        let mut result = Vec::with_capacity(nums.len() - k + 1);
        let mut max_indices = std::collections::VecDeque::with_capacity(k);
        max_indices.push_back(0 as usize);

        for end in 1..nums.len() {
            if max_indices.front().unwrap() + k - 1 < end {
                max_indices.pop_front();
            }
            while !max_indices.is_empty() && nums[*max_indices.back().unwrap()] < nums[end] {
                max_indices.pop_back();
            }
            max_indices.push_back(end);
            if end >= k - 1 {
                result.push(nums[*max_indices.front().unwrap()]);
            }
        }

        result
    }
}
